### PR TITLE
Optimizer support CTE plan

### PR DESCRIPTION
**Pull request by Seaven**
For #1318

### CODE REVIEW

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Review by dirtysalt — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/DeriveStatsTask.java [hunk 1]
```diff
@@ -44,9 +44,9 @@ public void execute() {
         }
 
         boolean needDerivedChildren = false;
-        // If we haven't got enough stats to compute the current stats, derive them
-        // from the child first
-        for (int i = 0; i < groupExpression.arity(); ++i) {
+        // If we haven't got enough stats to compute the current stats, derive them from the child first.
+        // For CTE, we need derive right tree first, then derive left
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/DeriveStatsTask.java:48]: derive right child first？

File: fe/fe-core/src/test/resources/sql/tpchcost/q20.sql [hunk 2]
```diff
@@ -333,12 +333,12 @@ column statistics:
     "test.nation": {
       "nation": 25
     },
-    "test.lineitem": {
-      "lineitem": 600000000
-    },
     "test.partsupp": {
       "partsupp": 80000000
     },
+    "test.lineitem": {
```
Comment (Youngwb) [inline fe/fe-core/src/test/resources/sql/tpchcost/q20.sql:339]: why these changed

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 3]
```diff
@@ -353,16 +355,6 @@ private void prepare() {
             fragmentExecParamsMap.put(fragment.getFragmentId(), new FragmentExecParams(fragment));
         }
 
-        // set inputFragments
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:356]: why remove this

File: fe/fe-core/src/main/java/com/starrocks/planner/MultiCastPlanFragment.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import com.clearspring.analytics.util.Lists;
+import com.google.common.base.Preconditions;
+import com.starrocks.analysis.Analyzer;
+import com.starrocks.common.Pair;
+import com.starrocks.sql.optimizer.statistics.ColumnDict;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class MultiCastPlanFragment extends PlanFragment {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/planner/MultiCastPlanFragment.java:14]: plz add some comment for this class

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/DeriveStatsTask.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/DeriveStatsTask.java:48]: It's work on stack, first input last out, so it's left child first

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:356]: Useless code

File: fe/fe-core/src/test/resources/sql/tpchcost/q20.sql (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/test/resources/sql/tpchcost/q20.sql:339]: I guess because I modify the useage of DumpLog

**Review by Youngwb — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**